core/src/main/java/com/scalar/db/transaction/consensuscommit/CommitHandler.java

```diff
@@ -50,7 +55,45 @@ protected void onPrepareFailure(Snapshot snapshot) {}
 
   protected void onValidateFailure(Snapshot snapshot) {}
 
+  private Optional<BeforePreparationSnapshotHookFuture> invokeBeforePreparationSnapshotHook(
+      Snapshot snapshot) throws UnknownTransactionStatusException, CommitException {
+    try {
+      if (beforePreparationSnapshotHook == null) {
+        return Optional.empty();
+      }
+
+      return Optional.of(beforePreparationSnapshotHook.handle(tableMetadataManager, snapshot));
+    } catch (Exception e) {
+      abortState(snapshot.getId());
+      rollbackRecords(snapshot);
+      // TODO: This method is actually a part of preparation phase. But the callback method name
+      //       `onPrepareFailure()` should be renamed to more reasonable one.
+      onPrepareFailure(snapshot);
+      throw new CommitException(e.getMessage(), e, snapshot.getId());
+    }
+  }
+
+  private void waitBeforePreparationSnapshotHookFuture(
+      Snapshot snapshot, @Nullable BeforePreparationSnapshotHookFuture snapshotHookFuture)
+      throws UnknownTransactionStatusException, CommitException {
+    if (snapshotHookFuture == null) {
+      return;
+    }
+    try {
+      snapshotHookFuture.get();
+    } catch (Exception e) {
+      abortState(snapshot.getId());
+      rollbackRecords(snapshot);
+      // TODO: This method is actually a part of validation phase. But the callback method name
+      //       `onValidateFailure()` should be renamed to more reasonable one.
+      onValidateFailure(snapshot);
+      throw new CommitException(e.getMessage(), e, snapshot.getId());
```

Comment:
Ditto.
```suggestion
      throw new CommitException(
          CoreError.WAITING_BEFORE_PREPARATION_SNAPSHOT_HOOK_FAILED.buildMessage(e.getMessage()),
          e,
          snapshot.getId());
```

Reply:
Good catch! I'll define and use the new error code.

Reply:
[Done!](https://github.com/scalar-labs/scalardb/pull/2271/commits/2e5db19cf6858c8dfae1843757832f109a198744)